File: sdks/go/integration_test.go
package allsource

import (
	"context"
	"fmt"
	"os"
	"testing"
	"time"
)

// Integration tests require ALLSOURCE_TEST_CORE_URL to be set.
// Run: ALLSOURCE_TEST_CORE_URL=http://localhost:3280 ALLSOURCE_TEST_API_KEY=test-key go test -v -run Integration

func coreURL(t *testing.T) string {
	t.Helper()
	url := os.Getenv("ALLSOURCE_TEST_CORE_URL")
	if url == "" {
		t.Skip("ALLSOURCE_TEST_CORE_URL not set, skipping integration test")
	}
	return url
}

func apiKey() string {
	key := os.Getenv("ALLSOURCE_TEST_API_KEY")
	if key == "" {
		return "test-key"
	}
	return key
}

func uniqueEntity(prefix string) string {
	return fmt.Sprintf("%s-%d", prefix, time.Now().UnixNano())
}

func TestIntegrationHealth(t *testing.T) {
	url := coreURL(t)
	client := New(apiKey(), url)

	result, err := client.Health(context.Background())
	if err != nil {
		t.Fatalf("health check failed: %v", err)
	}
	if result["status"] != "healthy" {
		t.Errorf("expected status healthy, got %v", result["status"])
	}
}

func TestIntegrationIngestAndQuery(t *testing.T) {
	url := coreURL(t)
	client := New(apiKey(), url)
	ctx := context.Background()
	entity := uniqueEntity("go-sdk-single")

	// Ingest
	resp, err := client.Ingest(ctx, "go.test.created", entity, map[string]any{
		"test":  true,
		"value": 42,
	})
	if err != nil {
		t.Fatalf("ingest failed: %v", err)
	}
	if resp.EventID == "" {
		t.Error("event_id should not be empty")
	}
	if resp.Timestamp == "" {
		t.Error("timestamp should not be empty")
	}

	// Query back
	result, err := client.Query(ctx, QueryOptions{EntityID: entity})
	if err != nil {
		t.Fatalf("query failed: %v", err)
	}
	if result.Count != 1 {
		t.Errorf("expected 1 event, got %d", result.Count)
	}
	if len(result.Events) != 1 {
		t.Fatalf("expected 1 event in array, got %d", len(result.Events))
	}
	if result.Events[0].EntityID != entity {
		t.Errorf("expected entity_id %s, got %s", entity, result.Events[0].EntityID)
	}
	if result.Events[0].EventType != "go.test.created" {
		t.Errorf("expected event_type go.test.created, got %s", result.Events[0].EventType)
	}
}

func TestIntegrationBatchIngest(t *testing.T) {
	url := coreURL(t)
	client := New(apiKey(), url)
	ctx := context.Background()
	entity := uniqueEntity("go-sdk-batch")

	events := make([]map[string]any, 5)
	for i := 0; i < 5; i++ {
		events[i] = map[string]any{
			"event_type": fmt.Sprintf("go.test.batch.item%d", i),
			"entity_id":  entity,
			"payload":    map[string]any{"index": i},
		}
	}

	resp, err := client.IngestBatch(ctx, events)
	if err != nil {
		t.Fatalf("batch ingest failed: %v", err)
	}
	if resp.Total != 5 {
		t.Errorf("expected total 5, got %d", resp.Total)
	}
	if resp.Ingested != 5 {
		t.Errorf("expected ingested 5, got %d", resp.Ingested)
	}
	if len(resp.Events) != 5 {
		t.Errorf("expected 5 event responses, got %d", len(resp.Events))
	}

	// Query back
	result, err := client.Query(ctx, QueryOptions{EntityID: entity})
	if err != nil {
		t.Fatalf("query failed: %v", err)
	}
	if result.Count != 5 {
		t.Errorf("expected 5 events, got %d", result.Count)
	}
}

func TestIntegrationQueryByType(t *testing.T) {
	url := coreURL(t)
	client := New(apiKey(), url)
	ctx := context.Background()
	entity := uniqueEntity("go-sdk-type")
	eventType := fmt.Sprintf("go.test.typed.%d", time.Now().UnixNano())

	for i := 0; i < 3; i++ {
		_, err := client.Ingest(ctx, eventType, entity, map[string]any{"seq": i})
		if err != nil {
			t.Fatalf("ingest %d failed: %v", i, err)
		}
	}

	result, err := client.Query(ctx, QueryOptions{EventType: eventType})
	if err != nil {
		t.Fatalf("query failed: %v", err)
	}
	if result.Count < 3 {
		t.Errorf("expected at least 3 events, got %d", result.Count)
	}
	for _, e := range result.Events {
		if e.EventType != eventType {
			t.Errorf("expected event_type %s, got %s", eventType, e.EventType)
		}
	}
}

func TestIntegrationQueryWithLimit(t *testing.T) {
	url := coreURL(t)
	client := New(apiKey(), url)
	ctx := context.Background()
	entity := uniqueEntity("go-sdk-limit")

	for i := 0; i < 5; i++ {
		_, err := client.Ingest(ctx, "go.test.limit", entity, map[string]any{"seq": i})
		if err != nil {
			t.Fatalf("ingest failed: %v", err)
		}
	}

	result, err := client.Query(ctx, QueryOptions{EntityID: entity, Limit: 2})
	if err != nil {
		t.Fatalf("query failed: %v", err)
	}
	if len(result.Events) != 2 {
		t.Errorf("expected 2 events with limit=2, got %d", len(result.Events))
	}
}

func TestIntegrationQueryNonexistent(t *testing.T) {
	url := coreURL(t)
	client := New(apiKey(), url)

	result, err := client.Query(context.Background(), QueryOptions{
		EntityID: "nonexistent-entity-go-sdk-99999",
	})
	if err != nil {
		t.Fatalf("query should succeed with empty: %v", err)
	}
	if result.Count != 0 {
		t.Errorf("expected 0 events, got %d", result.Count)
	}
}

func TestIntegrationPayloadFidelity(t *testing.T) {
	url := coreURL(t)
	client := New(apiKey(), url)
	ctx := context.Background()
	entity := uniqueEntity("go-sdk-fidelity")

	payload := map[string]any{
		"string":  "hello world",
		"number":  float64(42),
		"float":   3.14159,
		"boolean": true,
		"nested": map[string]any{
			"array": []any{float64(1), float64(2), float64(3)},
			"deep":  map[string]any{"key": "value"},
		},
	}

	_, err := client.Ingest(ctx, "go.test.fidelity", entity, payload)
	if err != nil {
		t.Fatalf("ingest failed: %v", err)
	}

	result, err := client.Query(ctx, QueryOptions{EntityID: entity})
	if err != nil {
		t.Fatalf("query failed: %v", err)
	}
	if result.Count != 1 {
		t.Fatalf("expected 1 event, got %d", result.Count)
	}

	stored := result.Events[0].Payload
	if stored["string"] != "hello world" {
		t.Errorf("expected string 'hello world', got %v", stored["string"])
	}
	if stored["number"] != float64(42) {
		t.Errorf("expected number 42, got %v", stored["number"])
	}
	if stored["boolean"] != true {
		t.Errorf("expected boolean true, got %v", stored["boolean"])
	}
	nested, ok := stored["nested"].(map[string]any)
	if !ok {
		t.Fatal("expected nested object")
	}
	deep, ok := nested["deep"].(map[string]any)
	if !ok {
		t.Fatal("expected deep object")
	}
	if deep["key"] != "value" {
		t.Errorf("expected deep.key=value, got %v", deep["key"])
	}
}

func TestIntegrationEventOrdering(t *testing.T) {
	url := coreURL(t)
	client := New(apiKey(), url)
	ctx := context.Background()
	entity := uniqueEntity("go-sdk-order")

	for i := 0; i < 5; i++ {
		_, err := client.Ingest(ctx, "go.test.ordered", entity, map[string]any{"seq": float64(i)})
		if err != nil {
			t.Fatalf("ingest failed: %v", err)
		}
		time.Sleep(10 * time.Millisecond)
	}

	result, err := client.Query(ctx, QueryOptions{EntityID: entity})
	if err != nil {
		t.Fatalf("query failed: %v", err)
	}
	if result.Count != 5 {
		t.Fatalf("expected 5 events, got %d", result.Count)
	}

	for i, e := range result.Events {
		seq, ok := e.Payload["seq"].(float64)
		if !ok {
			t.Fatalf("expected seq as float64 at index %d", i)
		}
		if int(seq) != i {
			t.Errorf("expected seq=%d at index %d, got %v", i, i, seq)
		}
	}
}

File: sdks/typescript/test/integration.test.ts
/**
 * Integration tests for the AllSource TypeScript SDK.
 *
 * Requires a running AllSource Core instance.
 *
 * Run:
 *   ALLSOURCE_TEST_CORE_URL=http://localhost:3280 \
 *   ALLSOURCE_TEST_API_KEY=test-key \
 *     bun test test/integration.test.ts
 */

import { describe, test, expect, beforeAll } from "bun:test";
import { AllSourceClient, AllSourceError } from "../src";

const CORE_URL = process.env.ALLSOURCE_TEST_CORE_URL;
const API_KEY = process.env.ALLSOURCE_TEST_API_KEY ?? "test-key";

function uniqueEntity(prefix: string): string {
  return `${prefix}-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`;
}

function skipUnlessCore() {
  if (!CORE_URL) {
    console.log(
      "SKIPPED: set ALLSOURCE_TEST_CORE_URL to run integration tests",
    );
    return true;
  }
  return false;
}

function makeClient(): AllSourceClient {
  return new AllSourceClient({ baseUrl: CORE_URL!, apiKey: API_KEY });
}

describe("Integration: Health", () => {
  test("core health returns healthy", async () => {
    if (skipUnlessCore()) return;
    const client = makeClient();
    const health = await client.getHealth();
    expect(health.status).toBe("healthy");
  });
});

describe("Integration: Single Event", () => {
  test("ingest and query round-trip", async () => {
    if (skipUnlessCore()) return;
    const client = makeClient();
    const entity = uniqueEntity("ts-sdk-single");

    // Ingest
    const resp = await client.ingestEvent({
      event_type: "ts.test.created",
      entity_id: entity,
      payload: { test: true, value: 42 },
      metadata: { source: "ts-sdk-integration-test" },
    });

    expect(resp.event_id).toBeTruthy();
    expect(resp.timestamp).toBeTruthy();

    // Query back
    const result = await client.queryEvents({ entity_id: entity });
    expect(result.count).toBe(1);
    expect(result.events.length).toBe(1);
    expect(result.events[0].entity_id).toBe(entity);
    expect(result.events[0].event_type).toBe("ts.test.created");
    expect(result.events[0].payload.value).toBe(42);
  });

  test("ingest with no metadata", async () => {
    if (skipUnlessCore()) return;
    const client = makeClient();
    const entity = uniqueEntity("ts-sdk-no-meta");

    const resp = await client.ingestEvent({
      event_type: "ts.test.minimal",
      entity_id: entity,
      payload: {},
    });

    expect(resp.event_id).toBeTruthy();
  });
});

describe("Integration: Batch Ingest", () => {
  test("batch ingest 5 events", async () => {
    if (skipUnlessCore()) return;
    const client = makeClient();
    const entity = uniqueEntity("ts-sdk-batch");

    const events = Array.from({ length: 5 }, (_, i) => ({
      event_type: `ts.test.batch.item${i}`,
      entity_id: entity,
      payload: { index: i },
    }));

    const resp = await client.ingestBatch(events);
    expect(resp.total).toBe(5);
    expect(resp.ingested).toBe(5);
    expect(resp.events.length).toBe(5);

    // Unique IDs
    const ids = new Set(resp.events.map((e) => e.event_id));
    expect(ids.size).toBe(5);

    // Query back
    const result = await client.queryEvents({ entity_id: entity });
    expect(result.count).toBe(5);
  });
});

describe("Integration: Query Filters", () => {
  test("query by event_type", async () => {
    if (skipUnlessCore()) return;
    const client = makeClient();
    const entity = uniqueEntity("ts-sdk-type");
    const eventType = `ts.test.typed.${Date.now()}`;

    for (let i = 0; i < 3; i++) {
      await client.ingestEvent({
        event_type: eventType,
        entity_id: entity,
        payload: { seq: i },
      });
    }

    const result = await client.queryEvents({ event_type: eventType });
    expect(result.count).toBeGreaterThanOrEqual(3);
    for (const event of result.events) {
      expect(event.event_type).toBe(eventType);
    }
  });

  test("query with limit", async () => {
    if (skipUnlessCore()) return;
    const client = makeClient();
    const entity = uniqueEntity("ts-sdk-limit");

    for (let i = 0; i < 5; i++) {
      await client.ingestEvent({
        event_type: "ts.test.limit",
        entity_id: entity,
        payload: { seq: i },
      });
    }

    const result = await client.queryEvents({
      entity_id: entity,
      limit: 2,
    });
    expect(result.events.length).toBe(2);
  });

  test("query nonexistent entity returns empty", async () => {
    if (skipUnlessCore()) return;
    const client = makeClient();
    const result = await client.queryEvents({
      entity_id: "nonexistent-ts-sdk-99999",
    });
    expect(result.count).toBe(0);
    expect(result.events.length).toBe(0);
  });
});

describe("Integration: Error Handling", () => {
  test("connection refused", async () => {
    const client = new AllSourceClient({
      baseUrl: "http://127.0.0.1:19999",
      apiKey: "key",
      timeout: 2000,
    });

    try {
      await client.getHealth();
      expect(true).toBe(false); // should not reach
    } catch (e) {
      expect(e).toBeDefined();
    }
  });
});

describe("Integration: Data Integrity", () => {
  test("payload fidelity", async () => {
    if (skipUnlessCore()) return;
    const client = makeClient();
    const entity = uniqueEntity("ts-sdk-fidelity");

    const complexPayload = {
      string: "hello world",
      number: 42,
      float: 3.14159,
      boolean: true,
      null_val: null,
      nested: {
        array: [1, 2, 3],
        deep: { key: "value" },
      },
    };

    await client.ingestEvent({
      event_type: "ts.test.fidelity",
      entity_id: entity,
      payload: complexPayload,
    });

    const result = await client.queryEvents({ entity_id: entity });
    expect(result.count).toBe(1);

    const stored = result.events[0].payload;
    expect(stored.string).toBe("hello world");
    expect(stored.number).toBe(42);
    expect(stored.boolean).toBe(true);
    expect(stored.null_val).toBeNull();
    expect((stored.nested as any).array[1]).toBe(2);
    expect((stored.nested as any).deep.key).toBe("value");
  });

  test("metadata preserved", async () => {
    if (skipUnlessCore()) return;
    const client = makeClient();
    const entity = uniqueEntity("ts-sdk-metadata");

    await client.ingestEvent({
      event_type: "ts.test.metadata",
      entity_id: entity,
      payload: { data: 1 },
      metadata: {
        correlation_id: "corr-123",
        source: "integration-test",
      },
    });

    const result = await client.queryEvents({ entity_id: entity });
    expect(result.count).toBe(1);
    expect(result.events[0].metadata.correlation_id).toBe("corr-123");
    expect(result.events[0].metadata.source).toBe("integration-test");
  });
});

describe("Integration: Event Ordering", () => {
  test("events returned in chronological order", async () => {
    if (skipUnlessCore()) return;
    const client = makeClient();
    const entity = uniqueEntity("ts-sdk-order");

    for (let i = 0; i < 5; i++) {
      await client.ingestEvent({
        event_type: "ts.test.ordered",
        entity_id: entity,
        payload: { seq: i },
      });
      await new Promise((r) => setTimeout(r, 10));
    }

    const result = await client.queryEvents({ entity_id: entity });
    expect(result.count).toBe(5);

    for (let i = 0; i < result.events.length; i++) {
      expect(result.events[i].payload.seq).toBe(i);
    }
  });
});
